_posts/2025-05-22-debezium-as-part-of-your-ai-solution.adoc

```diff
@@ -0,0 +1,440 @@
+
+=== How Debezium Milvus Sink Works
+
+The typical deployment of Milvus sink can be described with this workflow
+
+* Debezium Connector captures changes from the source database
+* If the source database does not contain a field with embedding then https://debezium.io/blog/2025/04/02/debezium-3-1-final-released/#new-features-and-improvements-ai[FieldToEmbedding SMT] is used to enrich change events with vector embeddings
+* Milvus sink writes the vector data into Milvus
+* Milvus stores and indexes the vectors for fast similarity search
+
+== Example application
+
+This blogpost is built around https://github.com/debezium/debezium-examples/tree/main/ai-rag[ai-rag] example application.
+We are going step-by-step through the details and internals of the application so you can easily built your own.
+
+image::overview.png[image,caption="Figure 1: Architecture",width=800,role=centered-image]
+
+The architecture identifies two different flows
+
+* Data pipeline
+* AI application
+
+that share common Milvus instance.
+Data pipeline is responsible for ingesting data into Milvus and AI application queries it.
+
+=== Data pipeline
+
+Our source of the documents is https://arxiv.org/[arXiv], service that provides scholarly articles for download in different formats.
+The test application downloads requested article based on its identifier and stores it in PostgreSQL table.
+
+The table is created with the following https://github.com/debezium/debezium-examples/blob/main/ai-rag/config-postgres/ai-sample-data.sql[DDL]
+[source,sql]
+----
+CREATE TABLE documents (
+  id VARCHAR(64) PRIMARY KEY,
+  metadata JSON,
+  text TEXT
+);
+ALTER TABLE documents REPLICA IDENTITY FULL;
+----
+
+The table has enabled full replica identity which means that original state of the database row will be available in update and delete messages.
+This is not required for our demo but can be useful for different implementations.
+
+Upon user request the application will download the document in text form and store it in the table.
+The document identifier will be used as the `id`, the textual representation of the article as `text` and JSON metadata `{"id:" "<document id>"}` in metadata.
+There is no column with embeddings so Debezium is responsible for calculation.
+
+We run LLM locally in Ollama LLM framework.
+As the model itself we selected https://ollama.com/library/granite3.1-dense:2b[granite3.1-dense:2b], that is text-only model optimized for also for RAG.
+It is one of large family of open-source https://huggingface.co/ibm-granite[IBM Granite models] but the application is not limited to it and can use an arbitrary one.
+
+Debezium Server links together PostgreSQL store with Milvus.
+As embeddings were not provided, they are calculated using https://github.com/debezium/debezium-examples/blob/main/ai-rag/config-server/application.properties#L16[FieldToEmbedding] SMT and stored in the field named `vector`.
```

Comment:
```suggestion
Because embeddings are not provided, we can use the https://github.com/debezium/debezium-examples/blob/main/ai-rag/config-server/application.properties#L16[Debezium FieldToEmbedding SMT] to calculate them and store them a field named `vector`.
```